monai/losses/focal_loss.py

```diff
@@ -57,80 +65,73 @@ def __init__(
+            target: the shape should be BNH[WD] or B1H[WD], where N is the number of classes.
 
         Raises:
-            ValueError: When ``target`` ndim differs from ``logits``.
-            ValueError: When ``target`` channel is not 1 and ``target`` shape differs from ``logits``.
+            AssertionError: When input and target (after one hot transform if setted)
+                have different shapes.
             ValueError: When ``self.reduction`` is not one of ["mean", "sum", "none"].
 
         """
-        i = logits
+        n_pred_ch = input.shape[1]
+
+        if self.to_onehot_y:
+            if n_pred_ch == 1:
+                warnings.warn("single channel prediction, `to_onehot_y=True` ignored.")
+            else:
+                target = one_hot(target, num_classes=n_pred_ch)
+
+        if not self.include_background:
+            if n_pred_ch == 1:
+                warnings.warn("single channel prediction, `include_background=False` ignored.")
+            else:
+                # if skipping background, removing first channel
+                target = target[:, 1:]
+                input = input[:, 1:]
+
+        if target.shape != input.shape:
+            raise AssertionError(f"ground truth has different shape ({target.shape}) from input ({input.shape})")
+
+        i = input
         t = target
 
-        if i.ndim != t.ndim:
-            raise ValueError(f"logits and target ndim must match, got logits={i.ndim} target={t.ndim}.")
-
-        if t.shape[1] != 1 and t.shape[1] != i.shape[1]:
-            raise ValueError(
-                "target must have one channel or have the same shape as the logits. "
-                "If it has one channel, it should be a class index in the range [0, C-1] "
-                f"where C is the number of classes inferred from 'logits': C={i.shape[1]}. "
-            )
-        if i.shape[1] == 1:
-            raise NotImplementedError("Single-channel predictions not supported.")
-
-        # Change the shape of logits and target to
-        # num_batch x num_class x num_voxels.
-        if i.dim() > 2:
-            i = i.view(i.size(0), i.size(1), -1)  # N,C,H,W => N,C,H*W
-            t = t.view(t.size(0), t.size(1), -1)  # N,1,H,W => N,1,H*W or N,C,H*W
-        else:  # Compatibility with classification.
-            i = i.unsqueeze(2)  # N,C => N,C,1
-            t = t.unsqueeze(2)  # N,1 => N,1,1 or N,C,1
-
-        # Compute the log proba (more stable numerically than softmax).
-        logpt = F.log_softmax(i, dim=1)  # N,C,H*W
-        # Keep only log proba values of the ground truth class for each voxel.
-        if target.shape[1] == 1:
-            logpt = logpt.gather(1, t.long())  # N,C,H*W => N,1,H*W
-            logpt = torch.squeeze(logpt, dim=1)  # N,1,H*W => N,H*W
+        # Change the shape of input and target to B x N x num_voxels.
+        i = i.view(i.size(0), i.size(1), -1)
+        t = t.view(t.size(0), t.size(1), -1)
```

Comment:
the previous implementation has the unsqueeze for classification, any reason for removing that?

Reply:
Hi @wyli , here I wish to unify the format between focal loss and other loss functions, thus I added `to_onehot_y` to ensure that the target will be changed into one-hot format and in this way, there is no need to handle the situation that `target.shape[1]` is 1.